source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc

```diff
@@ -21,10 +23,15 @@ LocalRateLimiterImpl::LocalRateLimiterImpl(
     throw EnvoyException("local rate limit token bucket fill timer must be >= 50ms");
   }
 
-  token_bucket_.max_tokens_ = max_tokens;
-  token_bucket_.tokens_per_fill_ = tokens_per_fill;
-  token_bucket_.fill_interval_ = absl::FromChrono(fill_interval);
-  tokens_.tokens_ = max_tokens;
+  RateLimit::TokenBucket default_token_bucket;
+  default_token_bucket.max_tokens_ = max_tokens;
+  default_token_bucket.tokens_per_fill_ = tokens_per_fill;
+  default_token_bucket.fill_interval_ = absl::FromChrono(fill_interval);
+  default_descriptor_.token_bucket_ = default_token_bucket;
+  auto default_token_state = std::make_unique<TokenState>();
+  default_token_state->tokens_ = max_tokens;
+  default_token_state->fill_time_ = time_source_.monotonicTime();
```

Comment:
Github prevent me from commenting on that `fill_time` thread :) So I re-raise it here. It is a good point that workaround doesn't work for http filter with no descriptor

Fundamentally, it is the issue in network filter test (not the cod here) that  `MockDispatcher` uses the GlobalTimeSystem and the `MockStreamInfo` in `MockReadFilterCallbacks` which uses the `SimulatedTimeSystem`

Have you tried to see if it works by switching default testSystem in class `GlobalTimeSystem` to `SimulatedTimeSystem`. Per TODO(#4160), it seems the direction to go anyway.

If that doesn't work, maybe we can pull in folks who have more experience on time system.

Reply:
I have performed a quick test by switching the time system in `MockDispatcher` from `GlobalTimeSystem` to `SimulatedTimeSystem`. After that, these tests passed with no errors.

In the README, a `GlobalTimeSystem` will initialize itself into `TestRealTimeSystem` but will be changed in the future to `SimulatedTimeSystem`, and this is a TODO in #4160.

So, shall we see and wait the change of the time system, or use the hacky way (in my commit) to switch the time system.

Reply:
The test results looks promising and the change is pretty self-contained in this test.  I feel it might depends on How long/hard it will take to switch default to `SimulatedTimeSystem` in `GlobalTimeSystem` vs How soon you want to check in this PR :)

I am not an expert on time system, let me pull in @jmarantz who filed the TODO #4160 to shed some lights here.
Josh, TL;DR, currently in network/local_ratelimit/local_ratelimit_test, there are two time system are used: `GlobalTimeSystem` (in MockDispatcher) and `SimulatedTimeSystem` (in MockStreamInfo). This issue is exposed by this PR. Per TODO, we know simulateTimeSystem seems to be direction to go. But in the scope of this PR, do you think it is OK that we just force `timeSystem` in the `MockDispatcher`  to `SimulatedTimeSystem` in this test as commit #a78997b pushed by Zhihao? Or do you have other insights on how to solve this issue? Thanks!

Reply:
I found it is not easy to switch to `SimulatedTimeSystem`, so maybe it is acceptable to go on?

Reply:
Sorry I missed this. in general we should only use one time system at a time, either a Simulated one or a Real one.

The "GlobalTimeSystem" is just a helper class to get access to a time-system reference to the Real or Simulated system that already has been declared (or to declare a default real-time system if one is not already declared).

To make a particular test use SimulatedTime you should only need to make its test-class derive from Event::TestUsingSimulatedTime, from test/test_common/simulated_time_system.h

Now -- you may experience issues with simulated time if there are parts of the test that depend on real time and don't provide a mechanism to inject time. One place I think this might come up with is gRPC which doesn't (afaik) provide a time-injection hook, but may use sleeps and real-time timeouts that don't work well with simulated time.

Reply:
https://github.com/envoyproxy/envoy/blob/a78997b3d4777b16b34740fd66f2bc7801719a42/test/extensions/filters/network/local_ratelimit/local_ratelimit_fuzz_test.cc#L52

Thanks for your reply. But we have a fuzz test which is also influenced by the time system. We cannot make the test class derive from `Event::TestUsingSimulatedTime` since there is no class in the fuzz test. Do you have any idea on how to solve that?